### PR TITLE
Refactor drawing

### PR DESCRIPTION
**Pull request by pakb**
Attempt to refactor most of the big DrawingModule and DrawingManager into smaller more manageable chunks

[Test link](https://web-mapviewer.dev.bgdi.ch/feature-jira-bgdiinf_sb-2242_refactor_drawing/index.html)

This is not meant to pass CI tests (will be done later in the `develop-refactor-drawing` branch, plus some other fixes

### CODE REVIEW

File: src/api/features.api.js [hunk 1]
```diff
@@ -17,72 +18,220 @@ export class Feature extends EventEmitter {
...
     }
-    set coordinate(newCoordinate) {
-        if (Array.isArray(newCoordinate) && newCoordinate.length === 2) {
-            this._coordinate = newCoordinate
+    set coordinates(newCoordinates) {
+        if (Array.isArray(newCoordinates)) {
+            // checking if we have received a unique coordinate
+            if (
+                newCoordinates.length === 2 &&
+                newCoordinates.filter((coord) => Array.isArray(coord)).length === 0
```
Comment (fkm) [inline src/api/features.api.js:64]:
```suggestion
                !newCoordinates.some((coord) => Array.isArray(coord))
```

File: src/api/features.api.js [hunk 2]
```diff
@@ -17,72 +18,220 @@ export class Feature extends EventEmitter {
...
+            ) {
+                // as we want an array of coordinates, we wrap the unique coordinate in an array
+                this._coordinates = [newCoordinates]
+            } else {
+                this._coordinates = newCoordinates
+            }
             this.emitChangeEvent()
```
Comment (fkm) [inline src/api/features.api.js:71]:
```suggestion
            this.emitChangeEvent('coordinates')
```

File: src/api/features.api.js [hunk 3]
```diff
@@ -17,72 +18,220 @@ export class Feature extends EventEmitter {
...
+     * Tells if the feature is currently being dragged (and later dropped) by the user
+     *
+     * @returns {boolean}
+     */
+    get isDragged() {
+        return this._isDragged
+    }
+    set isDragged(flag) {
+        this._isDragged = flag
+        this.emitChangeEvent('isDragged')
```
Comment (fkm) [inline src/api/features.api.js:109]: Shouldn't this only be emitted if `flag === true`?

File: src/api/features.api.js [hunk 4]
```diff
@@ -17,72 +18,220 @@ export class Feature extends EventEmitter {
...
+    }
+    /** @param newSize {FeatureStyleSize} */
+    set textSize(newSize) {
+        if (newSize && allStylingSizes.find((size) => size.textScale === newSize.textScale)) {
+            this._textSize = newSize
+            this.emitChangeEvent('textSize')
+        }
+    }
+    get font() {
+        return this._textSize.font
```
Comment (fkm) [inline src/api/features.api.js:192]:
```suggestion
        return this._textSize?.font
```

File: src/api/features.api.js [hunk 5]
```diff
@@ -93,29 +242,46 @@ export class LayerFeature extends Feature {
...
+
+    // getters for all attributes (no setters)
+    get layer() {
+        return this._layer
     }
 
     /** @returns {LayerTypes} */
     getLayerType() {
-        return this.layer && this.layer.type
+        return this._layer && this._layer.type
```
Comment (fkm) [inline src/api/features.api.js:272]:
```suggestion
        return this._layer?.type
```

File: src/api/icon.api.js [hunk 6]
```diff
@@ -1,42 +1,142 @@
...
+        return this._imageTemplateURL
+            .replace('{icon_set_name}', this._iconSetName)
+            .replace('{icon_name}', this._name)
             .replace('{icon_scale}', iconSize.iconScale + 'x')
-            .replace('{r}', rgb[0])
-            .replace('{g}', rgb[1])
-            .replace('{b}', rgb[2])
+            .replace('{r}', `${rgb[0]}`)
+            .replace('{g}', `${rgb[1]}`)
+            .replace('{b}', `${rgb[2]}`)
```
Comment (fkm) [inline src/api/icon.api.js:139]: I'm curious. Did this pose a problem? (when passed directly)

**Review by fkm — COMMENTED** [review]
I know it's still a draft. But I couldn't help myself and made a first pass :see_no_evil: 

It's all just suggestions/questions! You can resolve the comments as you see fit.

File: src/api/features.api.js (hunk 3 above)
Comment (pakb) [reply src/api/features.api.js:109]: mmh, yes and no, it means the value of `isDragged` has changed (an event called `change:isDragged` will be fired), maybe the naming is not good enough.

File: src/api/icon.api.js (hunk 6 above)
Comment (pakb) [reply src/api/icon.api.js:139]: My IDE was complaining about the unmatching types (JS will do the conversion behind the scene anyway...)
I don't remember if the linter was also complaining though

File: src/api/features.api.js (hunk 3 above)
Comment (fkm) [reply src/api/features.api.js:109]: Ok. I think I get it. So, we' use it like this?
`.on('isDragged', (feature) => if (feature.isDragged) { ... })`

File: src/api/icon.api.js (hunk 6 above)
Comment (fkm) [reply src/api/icon.api.js:139]: Ok :smiley:

File: src/api/features.api.js (hunk 3 above)
Comment (pakb) [reply src/api/features.api.js:109]: yep, that's the idea

Comment (fkm) [reply src/api/features.api.js:109]: :+1:

**Review by fkm — COMMENTED** [review]
I have 8 files to go. But the last few files will take a while.

Here's what I got so far.

File: src/modules/drawing/components/DrawingTooltip.vue [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+            default: null,
+        },
+        selectedFeatures: {
+            type: Array,
+            default: () => [],
+        },
+    },
+    data() {
+        return {
+            sketchPoints: 0,
```
[the remark this answers is not in the document]
Comment (pakb) [reply src/modules/drawing/components/DrawingTooltip.vue:23]: some artifacts from old code, a line points counter, I'll remove it (this is now handled by `drawLineMixin`)

File: src/modules/drawing/components/DrawingTooltip.vue [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    },
+    mounted() {
+        this.tooltipOverlay.setElement(this.$refs.drawingTooltip)
+        const map = this.getMap()
+        if (map) {
+            map.addOverlay(this.tooltipOverlay)
+            map.on('pointermove', this.onPointerMove)
+        }
+    },
+    beforeUnmount() {
```
[the remark this answers is not in the document]
Comment (pakb) [reply src/modules/drawing/components/DrawingTooltip.vue:43]: I think it's a good idea to reverse what we do in `mount()`, let's add that

File: src/modules/drawing/components/DrawingTooltip.vue [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        const map = this.getMap()
+        if (map) {
+            map.removeOverlay(this.tooltipOverlay)
+            map.un('pointermove', this.onPointerMove)
+        }
+    },
+    methods: {
+        updateTooltipText(translationKeyLine1, translationKeyLine2 = null) {
```
[the remark this answers is not in the document]
Comment (pakb) [reply src/modules/drawing/components/DrawingTooltip.vue:52]: I've not really finished working on this component yet, I've only moved the code that was doing the tooltip in it and commented all that was not working out of the box.
So maybe we can do what you are suggesting while finishing all the features of this component.

[the remark this answers is not in the document]
Comment (fkm) [reply src/modules/drawing/components/DrawingTooltip.vue:52]: Ok. Sounds good!

**Review by fkm — APPROVED** [review]

**Pull request merged**